chalk-solve/src/clauses/program_clauses.rs

```diff
@@ -180,6 +180,95 @@ impl<I: Interner> ToProgramClauses<I> for OpaqueTyDatum<I> {
     }
 }
 
+fn application_ty<I: Interner>(
+    builder: &mut ClauseBuilder<'_, I>,
+    id: impl CastTo<TypeName<I>>,
+) -> ApplicationTy<I> {
+    let interner = builder.interner();
+    ApplicationTy {
+        name: id.cast(interner),
+        substitution: builder.substitution_in_scope(),
+    }
+}
+
+fn interned_application_ty<I: Interner>(interner: &I, appl_ty: &ApplicationTy<I>) -> Ty<I> {
+    appl_ty.clone().intern(interner)
+}
+
+fn well_formed_program_clauses<'a, I, Wc>(
+    builder: &'a mut ClauseBuilder<'_, I>,
+    id: impl CastTo<TypeName<I>>,
+    where_clauses: Wc,
+) where
+    I: Interner,
+    Wc: Iterator<Item = &'a QuantifiedWhereClause<I>>,
+{
+    // Given a type definition `struct Foo<T: Eq> { }` generate the clause
+    // forall<T> {
+    //     WF(Foo<T>) :- WF(T: Eq).
+    // }
+    let interner = builder.interner();
+    let appl_ty = application_ty(builder, id);
+    let ty = appl_ty.clone().intern(interner);
+    builder.push_clause(
+        WellFormed::Ty(ty.clone()),
+        where_clauses
+            .cloned()
+            .map(|qwc| qwc.into_well_formed_goal(interner)),
+    );
+}
+
+fn fully_visible_program_clauses<'a, I>(
+    builder: &'a mut ClauseBuilder<'_, I>,
+    id: impl CastTo<TypeName<I>>,
+) where
+    I: Interner,
+{
+    // Given a type definition `struct Foo<T: Eq> { }` generate the clause
+    // forall<T> {
+    //     IsFullyVisible(Foo<T>) :- IsFullyVisible(T).
+    // }
+    let interner = builder.interner();
+    let appl_ty = application_ty(builder, id);
+    let ty = interned_application_ty(interner, &appl_ty);
+    builder.push_clause(
+        DomainGoal::IsFullyVisible(ty.clone()),
+        appl_ty
+            .type_parameters(interner)
+            .map(|typ| DomainGoal::IsFullyVisible(typ).cast::<Goal<_>>(interner)),
+    );
+}
+
+fn implied_bounds_program_clauses<'a, I, Wc>(
```

Comment:
This could use a comment. Something like:

/// Generates the "implied bounds" clauses for an applicative 
/// type with the name `type_name`. As an example, if `type_name`
/// represents a struct `S` that is declared like:
///
/// ```rust
/// struct S<T> where T: Eq { .. }
/// ```
///
/// then we would generate the rule
///
/// ```notrust
/// FromEnv(T: Eq) :- FromEnv(S<T>)
/// ```
///
/// # Parameters
///
/// - `builder` -- the clause builder. We assume all the generic types from `S` are in scope.
/// - `type_name` -- in our example above, the name `S`
/// - `where_clauses` -- the list of where clauses declared on the type (`T: Eq`, in our example).